File: second_year/projects/G-OOP-400-COT-4-1-bsraytracer-13/src/point3D.cpp
/*
** EPITECH PROJECT, 2026
** snd_year_projects
** File description:
** point3D
*/

#include "../include/point3D.hpp"
using namespace Math;

Point3D::Point3D(double valX, double valY, double valZ)
    : X(valX), Y(valY), Z(valZ)
{}

Point3D Point3D::operator+(const Vector3D& vector)
{
    return Point3D(this->X + vector.X, this->Y + vector.Y, this->Z + vector.Z);
}

Point3D Point3D::operator+(const Vector3D& vector) const
{
    return Point3D(this->X + vector.X, this->Y + vector.Y, this->Z + vector.Z);
}

Point3D Point3D::operator+=(const Vector3D& vector)
{
    Point3D _new = *this + vector;
    *this = _new;
    return *this;
}

Vector3D Point3D::operator-(const Point3D& other)
{
    return Vector3D(this->X - other.X, this->Y - other.Y, this->Z - other.Z);
}

Vector3D Point3D::operator-(const Point3D& other) const
{
    return Vector3D(this->X - other.X, this->Y - other.Y, this->Z - other.Z);
}

File: second_year/projects/G-OOP-400-COT-4-1-bsraytracer-13/rust/src/point3d.rs
//
// EPITECH PROJECT, 2026
// snd_year_projects
// File description:
// point3D
//

use crate::Vector3D;
use std::ops;

#[derive(Debug, Clone, Copy)]
pub struct Point3D {
    pub x: f32,
    pub y: f32,
    pub z: f32,
}

impl ops::Add<Vector3D> for Point3D {
    type Output = Point3D;

    fn add(self, vector: Vector3D) -> Point3D {
        return Point3D { x: (self.x + vector.x), y: (self.y + vector.y), z: (self.z + vector.z) };
    }
}

impl ops::AddAssign<Vector3D> for Point3D {
    
    fn add_assign(&mut self, vector: Vector3D) {
        let result = *self + vector;
        *self = result;
    }
}

impl ops::Sub<Point3D> for Point3D {
    type Output = Vector3D;

    fn sub(self, other: Point3D) -> Vector3D {
        return Vector3D { x: (self.x - other.x), y: (self.y - other.y), z: (self.z - other.z) };
    }
}
